### ISSUE TITLE
Docker Internal DNS Issue

### ISSUE THREAD

**Issue by elibrody** [louislam/uptime-kuma#4265]
### ⚠️ Please verify that this bug has NOT been raised before.

- [X] I checked and didn't find similar issue

### 🛡️ Security Policy

- [X] I agree to have read this project [Security Policy](https://github.com/louislam/uptime-kuma/security/policy)

### 📝 Describe your problem

Im having some weird dns resolution issues in the docker container pretty much no internet at all in the container any way to just force UptimeKuma to use 1.1.1.1 for dns instead of the Docker DNS?

### 🐻 Uptime-Kuma Version

1.23.10

### 💻 Operating System and Arch

Debian 12 x86

### 🌐 Browser

Google Chrome (Latest)

### 🐋 Docker Version

24.0.7

**Comment by CommanderStorm** [comment]
Please refer to your runtimes documentation for dns configuration as we use the one provided by the runtime.

https://docs.docker.com/network/#dns-services

**Comment by elibrody** [comment]
> Please refer to your runtimes documentation for how to configure cloudlare as a dns resolver as we use the one provided by the runtime.
> 
> https://docs.docker.com/network/#dns-services

Hey sorry but can you please elaborate what you mean by runtime?

**Comment by elibrody** [comment]
found some logs doesn't give me much but im sharing all it told me was that it is using the docker internal dns instead of the DNS provided in the docker compose 

2023-12-20T23:09:19Z ERR edge discovery: error looking up Cloudflare edge IPs: the DNS query failed error="lookup _v2-origintunneld._tcp.argotunnel.com on 127.0.0.11:53: read udp 127.0.0.1:43458->127.0.0.11:53: i/o timeout" event=0
2023-12-20T23:09:19Z ERR Please try the following things to diagnose this issue: event=0
2023-12-20T23:09:19Z ERR   1. ensure that argotunnel.com is returning "origintunneld" service records. event=0
2023-12-20T23:09:19Z ERR      Run your system's equivalent of: dig srv _origintunneld._tcp.argotunnel.com event=0
2023-12-20T23:09:19Z ERR   2. ensure that your DNS resolver is not returning compressed SRV records. event=0
2023-12-20T23:09:19Z ERR      See GitHub issue https://github.com/golang/go/issues/27546 event=0
2023-12-20T23:09:19Z ERR      For example, you could use Cloudflare's 1.1.1.1 as your resolver: event=0
2023-12-20T23:09:19Z ERR      https://developers.cloudflare.com/1.1.1.1/setting-up-1.1.1.1/ event=0
2023-12-20T23:09:19Z INF Tunnel server stopped
2023-12-20T23:09:19Z ERR Initiating shutdown error="Could not lookup srv records on _v2-origintunneld._tcp.argotunnel.com: lookup _v2-origintunneld._tcp.argotunnel.com on 127.0.0.11:53: read udp 127.0.0.1:43458->127.0.0.11:53: i/o timeout"
2023-12-20T23:09:19Z INF Metrics server stopped
Could not lookup srv records on _v2-origintunneld._tcp.argotunnel.com: lookup _v2-origintunneld._tcp.argotunnel.com on 127.0.0.11:53: read udp 127.0.0.1:43458->127.0.0.11:53: i/o timeout

**Comment by CommanderStorm** [comment]
- > can you please elaborate what you mean by runtime
  
  I am referig to the [docker runtime](https://www.docker.com/products/container-runtime/), i.e. the thing that executes the work created by `docker run`
  
  Adding the dns configuration to your `docker run` command might fix the issue you are experiencing.
- Please post a full configuration of your docker configuration (are you using an user defined docker network by any chance?) and network setup (you seem to run a cloudflare tunnel, correct?). Are you using custom dns? Is dns available on the "real" host outside the container?
- Just so we are on the same page: when you say your container does not have dns, is this reproducible using the steps in our [troubleshooting guide](https://github.com/louislam/uptime-kuma/wiki/Troubleshooting)?

**Comment by elibrody** [comment]
> * > can you please elaborate what you mean by runtime
>   
>   
>   I am referig to the [docker runtime](https://www.docker.com/products/container-runtime/), i.e. the thing that executes the work created by `docker run`
>   Adding the dns configuration to your `docker run` command might fix the issue you are experiencing.
> * Please post a full configuration of your docker configuration (are you using an user defined docker network by any chance?) and network setup (you seem to run a cloudflare tunnel, correct?). Are you using custom dns? Is dns available on the "real" host outside the container?
> * Just so we are on the same page: when you say your container does not have dns, is this reproducible using the steps in our [troubleshooting guide](https://github.com/louislam/uptime-kuma/wiki/Troubleshooting)?

I deployed in a vultr Debian 12 installed latest docker and it was running fine till yesterday see my docker compose (the port is changed since its exposed via Cloudflare tunnel 

in regards to the troubleshooting guide thats exactly what I followed to confirm the dns issue is in the container 

`version: '3.8'

services:
  uptime-kuma:
    image: louislam/uptime-kuma:1
    container_name: uptime-kuma
    volumes:
      - uptime-kuma:/app/data
    ports:
      - "3001"
    restart: always

volumes:
  uptime-kuma:
`

**Comment by chakflying** [comment]
For docker compose, you can set the DNS server used for a specific service:

https://docs.docker.com/compose/compose-file/05-services/#dns

**Comment by elibrody** [comment]
> For docker compose, you can set the DNS server used for a specific service:
> 
> https://docs.docker.com/compose/compose-file/05-services/#dns

if only that would work see this issue linked
[](https://github.com/docker/compose/issues/9550)

**Comment by chakflying** [comment]
If you think your issue is related to that, you should try the solutions proposed in that post (use the classic builder or avoid `compose` by using `docker run ...`).

**Comment by elibrody** [comment]
> If you think your issue is related to that, you should try the solutions proposed in that post (use the classic builder or avoid `compose` by using `docker run ...`).

So just tried using docker run was able to force a DNS nameserver change but still no internet in the container

**Comment by CommanderStorm** [comment]
- > but still no internet in the container
  
  Just that we are on the same page: With no internet you still mean no dns resolution, correct? Please post what you are doing more exactly (otherwise we might miss-communicate, text is very limiting)
- > Are you using custom dns? Is dns available on the "real" host outside the container?

  Let me rephrase: Does the same happen outside of docker? Might this be the VM which is at fault instead of the container?

**Comment by elibrody** [comment]
> * > but still no internet in the container
>   
>   
>   Just that we are on the same page: With no internet you still mean no dns resolution, correct? Please post what you are doing more exactly (otherwise we might miss-communicate, text is very limiting)

no network at all pinging 8.8.8.8 is dead too

> * > Are you using custom dns? Is dns available on the "real" host outside the container?
>   
DNS is available on the host just fine 
>   
>   Let me rephrase: Does the same happen outside of docker? Might this be the VM which is at fault instead of the container?
> * So that we are on the same page: how are you verifying that dns is not working?

https://github.com/louislam/uptime-kuma/wiki/Troubleshooting

**Comment by CommanderStorm** [comment]
> no network at all pinging 8.8.8.8 is dead too

So your symptom is not "no DNS" but rather "cannot ping any IP outside of the container".
Have you tried something like `dig google.com` to check if dns does not work?

This issue seems quite similar to this one: https://www.reddit.com/r/docker/comments/13mfc7f/docker_networking_seems_to_have_completely_broken/

=> could you try if your firewall/… is a culprit?

**Comment by elibrody** [comment]
Hi all this is resolved now by installing this 

https://github.com/chaifeng/ufw-docker?tab=readme-ov-file

Thanks all for the help
